File: tests/oracle/printc_terminal_1204.cc
#include "printc.hh"
#include "funcdata.hh"

#include <algorithm>
#include <iomanip>
#include <iostream>
#include <list>
#include <sstream>
#include <string>
#include <vector>

namespace {

using namespace ghidra;

// Access control is the only obstacle to constructing a minimal BlockBasic.
// Explicit template instantiation ignores member access control; the invoked
// method is the real locked BlockBasic::insert and preserves parent/order state.
struct BlockInsertTag {
  using type = void (BlockBasic::*)(std::list<PcodeOp *>::iterator, PcodeOp *);
  friend type access(BlockInsertTag);
};

struct OpSetInputTag {
  using type = void (PcodeOp::*)(Varnode *, int4);
  friend type access(OpSetInputTag);
};

template <typename Tag, typename Tag::type Member>
struct PrivateAccess {
  friend typename Tag::type access(Tag) { return Member; }
};

template struct PrivateAccess<BlockInsertTag, &BlockBasic::insert>;
template struct PrivateAccess<OpSetInputTag, &PcodeOp::setInput>;

class FixturePrintC final : public PrintC {
public:
  FixturePrintC(void) : PrintC(nullptr, "printc-terminal-1204") {
    delete emit;
    emit = new EmitNoMarkup();
  }

  void render(const BlockBasic *block, bool no_branch_active) {
    if (no_branch_active)
      setMod(no_branch);
    else
      unsetMod(no_branch);
    emitBlockBasic(block);
    emit->flush();
  }
};

void append(BlockBasic &block, PcodeOp *op) {
  const auto insert = access(BlockInsertTag{});
  (block.*insert)(block.endOp(), op);
}

std::string render(const std::vector<OpCode> &opcodes, bool no_branch_active) {
  PcodeOpBank bank;
  BlockBasic block(nullptr);
  TypeOpBranch branch_type(nullptr);
  TypeOpCbranch cbranch_type(nullptr);
  TypeOpReturn return_type(nullptr);
  ConstantSpace constant_space(nullptr, nullptr);
  TypeBase bool_type(1, TYPE_BOOL);
  Varnode target(8, Address(&constant_space, 0x1000), &bool_type);
  Varnode condition(1, Address(&constant_space, 1), &bool_type);
  new HighVariable(&target);
  new HighVariable(&condition);

  for (const OpCode opcode : opcodes) {
    TypeOp *type = nullptr;
    int4 input_count = 0;
    switch (opcode) {
    case CPUI_BRANCH:
      type = &branch_type;
      break;
    case CPUI_CBRANCH:
      type = &cbranch_type;
      input_count = 2;
      break;
    case CPUI_RETURN:
      type = &return_type;
      break;
    default:
      throw LowlevelError("unsupported fixture opcode");
    }
    PcodeOp *op = bank.create(input_count, Address());
    bank.changeOpcode(op, type);
    if (opcode == CPUI_CBRANCH) {
      const auto set_input = access(OpSetInputTag{});
      (op->*set_input)(&target, 0);
      (op->*set_input)(&condition, 1);
    }
    append(block, op);
  }

  std::ostringstream output;
  FixturePrintC printer;
  printer.setOutputStream(&output);
  printer.render(&block, no_branch_active);
  return output.str();
}

std::string toHex(const std::string &value) {
  std::ostringstream stream;
  stream << std::hex << std::setfill('0');
  for (const unsigned char byte : value)
    stream << std::setw(2) << static_cast<unsigned int>(byte);
  return stream.str();
}

void emitCase(const std::string &name, const std::vector<OpCode> &opcodes,
              bool no_branch_active, bool raw) {
  const std::string output = render(opcodes, no_branch_active);
  std::cout << name << '=';
  if (raw)
    std::cout << toHex(output);
  else
    std::cout << std::count(output.begin(), output.end(), ';');
  std::cout << '\n';
}

} // namespace

int main(int argc, char **argv) {
  const bool raw = argc == 2 && std::string(argv[1]) == "--raw";
  emitCase("cbranch_visible", {CPUI_CBRANCH}, false, raw);
  emitCase("cbranch_suppressed", {CPUI_CBRANCH}, true, raw);
  emitCase("branch_visible", {CPUI_BRANCH}, false, raw);
  emitCase("return_suppressed", {CPUI_RETURN}, true, raw);
  emitCase("mixed_visible", {CPUI_RETURN, CPUI_CBRANCH}, false, raw);
  emitCase("mixed_suppressed", {CPUI_RETURN, CPUI_CBRANCH}, true, raw);
  return 0;
}

File: tests/oracle/printc_terminal_1204.rs
use rugra::address::{Address, SeqNum};
use rugra::op::{PcodeOp, PcodeOpRef};
use rugra::opcodes::OpCode;
use rugra::prettyprint::EmitNoMarkup;
use rugra::printc::PrintC;
use rugra::space::AddressSpace;
use rugra::varnode::Varnode;
use std::sync::{Arc, RwLock};

fn make_op(opcode: OpCode, index: u32) -> PcodeOpRef {
    let mut op = PcodeOp::new(SeqNum::new(Address::new(index as u64), index), opcode);
    op.set_opcode_flags(opcode);
    if opcode == OpCode::CPUI_CBRANCH {
        op.inrefs.push(Arc::new(RwLock::new(Varnode::new_with_space(
            8,
            AddressSpace::Const,
            0x1000,
        ))));
        op.inrefs.push(Arc::new(RwLock::new(Varnode::new_with_space(
            1,
            AddressSpace::Const,
            1,
        ))));
    }
    PcodeOpRef(Arc::new(RwLock::new(op)))
}

fn render(opcodes: &[OpCode], suppress_branch: bool) -> String {
    let ops = opcodes
        .iter()
        .enumerate()
        .map(|(index, opcode)| make_op(*opcode, index as u32))
        .collect::<Vec<_>>();
    let mut printer = PrintC::new(Box::new(EmitNoMarkup::new()));
    printer.emit_block_basic_rpn(&ops, suppress_branch);
    let emit = printer
        .take_emit()
        .into_any()
        .downcast::<EmitNoMarkup>()
        .expect("PrintC fixture must retain EmitNoMarkup");
    emit.debug_get_output_ref().to_owned()
}

fn to_hex(value: &str) -> String {
    value
        .as_bytes()
        .iter()
        .map(|byte| format!("{byte:02x}"))
        .collect()
}

fn emit_case(name: &str, opcodes: &[OpCode], suppress_branch: bool, raw: bool) {
    let output = render(opcodes, suppress_branch);
    if raw {
        println!("{name}={}", to_hex(&output));
    } else {
        println!("{name}={}", output.matches(';').count());
    }
}

fn main() {
    let raw = std::env::args().nth(1).as_deref() == Some("--raw");
    emit_case("cbranch_visible", &[OpCode::CPUI_CBRANCH], false, raw);
    emit_case("cbranch_suppressed", &[OpCode::CPUI_CBRANCH], true, raw);
    emit_case("branch_visible", &[OpCode::CPUI_BRANCH], false, raw);
    emit_case("return_suppressed", &[OpCode::CPUI_RETURN], true, raw);
    emit_case(
        "mixed_visible",
        &[OpCode::CPUI_RETURN, OpCode::CPUI_CBRANCH],
        false,
        raw,
    );
    emit_case(
        "mixed_suppressed",
        &[OpCode::CPUI_RETURN, OpCode::CPUI_CBRANCH],
        true,
        raw,
    );
}
